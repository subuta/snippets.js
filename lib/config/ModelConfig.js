import check from 'lib/_utils/apiCheck'

const ModelConfigShape = check.shape({
  schema: check.object
})

// eg:
// config = {
//   book: {
//     schema: {}
//   }
// }
export const validateModels = function (config) {
  return check.throw(check.objectOf(ModelConfigShape), config)
}

// eg:
// config = {
//   schema: {}
// }
export const validateModel = function (config) {
  return check.throw(ModelConfigShape, config)
}
