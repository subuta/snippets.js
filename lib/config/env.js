import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'

export default (imports = []) => {
  const vars = _.zipObject(imports, imports)

  return build`
    // load .env first.
    require('dotenv').config()
    
    const ${s.raw(vars)} = process.env
    
    export ${s.raw(vars)}
    
    export default ${s.raw(vars)}
  `
}
