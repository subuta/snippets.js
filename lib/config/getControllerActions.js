import _ from 'lodash'

export default (config = {}) => {
  const {
    only, // the routes will generated.
    except  // the routes will ignored.
  } = config

  const routes = [
    'index',
    'show',
    'create',
    'update',
    'destroy'
  ]

  if (only) {
    return _.intersection(routes, only)
  } else if (except) {
    return _.difference(routes, except)
  }

  return routes
}
