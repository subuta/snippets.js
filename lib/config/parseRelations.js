import _ from 'lodash'
import pluralize from 'pluralize'

export default (config = {}) => {
  const { schema } = config

  const relations = _.transform(schema.relations, (result, value, key) => {
    let relatedModel = null;
    if (value.belongsTo) {
      relatedModel = pluralize.singular(_.toLower(value.belongsTo))
    } else if (value.hasOne) {
      relatedModel = pluralize.singular(_.toLower(value.hasOne))
    } else if (value.hasMany) {
      relatedModel = pluralize(_.toLower(value.hasMany))
    } else if (value.hasAndBelongsToMany) {
      relatedModel = pluralize(_.toLower(value.hasAndBelongsToMany))
    }
    result[key] = relatedModel
  }, {})

  return relations
}
