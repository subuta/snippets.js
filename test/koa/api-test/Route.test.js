/* global expect, describe, it, beforeEach, afterEach */

import { build, format, snippets as s } from 'bld.js'

import RouteTest from 'lib/koa/api-test/Route'

import { Routes as RoutesConfig, Models as ModelsConfig } from 'test/fixtures/config'

describe('koa/api-test/Route', () => {
  beforeEach(() => {
  })

  afterEach(() => {
  })

  it('should create koa Route', () => {
    expect(format(RouteTest({
      model: 'user',
      routeConfig: RoutesConfig.user,
      modelConfig: ModelsConfig.user
    }))).toMatchSnapshot()
  })
})
