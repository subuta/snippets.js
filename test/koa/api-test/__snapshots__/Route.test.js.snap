// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`koa/api-test/Route should create koa Route 1`] = `
"import test from 'ava'
import _ from 'lodash'
import sinon from 'sinon'
import request from 'supertest'
import {jwksEndpoint} from 'jwks-rsa/tests/mocks/jwks'
import {publicKey, privateKey} from 'jwks-rsa/tests/mocks/keys'
import {createToken} from 'jwks-rsa/tests/mocks/tokens'
import Koa from 'koa'
import importFresh from 'import-fresh'
import {absolutePath} from '../../config'
import {currentUser} from 'test/helper/user'
import runSeed, {runMigration} from 'test/helper/fixtures'
import proxyquire from 'proxyquire'

const sandbox = sinon.sandbox.create()

test.beforeEach(async (t) => {
  const knex = importFresh(absolutePath('src/utils/knex')).default

  await runMigration(knex)
  await runSeed(knex)

  const api = require('test/helper/mocked').api(knex)
  const models = require('test/helper/mocked').model(knex)

  const app = new Koa()
  // handle /api requests
  app.use(api.routes())
  app.use(api.allowedMethods())

  t.context = {
    ...models,
    request: request(app.listen(0))
  }
})

test.afterEach((t) => {
  sandbox.reset()
})

test('index should list user', async (t) => {
  const {request} = t.context

  // mock jwks
  const token = createToken(privateKey, '123', currentUser)
  jwksEndpoint('http://localhost', [{pub: publicKey, kid: '123'}])

  const response = await request
    .get('/api/users')
    .set('Authorization', \`Bearer \${token}\`)

  t.is(response.status, 200)
  t.deepEqual(response.body.length, 3)
  t.deepEqual(response.body, [19270, 82970, 93464])
})
test('show should return user', async (t) => {
  const {request} = t.context

  // mock jwks
  const token = createToken(privateKey, '123', currentUser)
  jwksEndpoint('http://localhost', [{pub: publicKey, kid: '123'}])

  const response = await request
    .get('/api/users/19270')
    .set('Authorization', \`Bearer \${token}\`)

  t.is(response.status, 200)

  t.deepEqual(response.body.id, 19270)
  t.deepEqual(response.body.auth0Id, 79736)
  t.deepEqual(response.body.locale, 'en_GB')
  t.deepEqual(response.body.nickname, 'Investment Account XSS')
  t.deepEqual(response.body.status, 'Fresh enhance')
  t.deepEqual(
    response.body.avatar,
    'https://s3.amazonaws.com/uifaces/faces/twitter/buzzusborne/128.jpg'
  )
  t.deepEqual(response.body.created_at, '2000-09-25T18:59:42.227Z')
  t.deepEqual(response.body.updated_at, '2000-09-15T04:49:30.128Z')
})
test('post should create user', async (t) => {
  const {request} = t.context

  // mock jwks
  const token = createToken(privateKey, '123', currentUser)
  jwksEndpoint('http://localhost', [{pub: publicKey, kid: '123'}])

  const response = await request
    .post('/api/users')
    .set('Authorization', \`Bearer \${token}\`)
    .send({
      article: {
        id: 90001,
        auth0Id: 60926,
        locale: 'en_BORK',
        nickname: 'neural',
        status: 'magenta Personal Loan Account',
        avatar:
          'https://s3.amazonaws.com/uifaces/faces/twitter/koridhandy/128.jpg',
        created_at: '2000-05-02T18:39:05.087Z',
        updated_at: '2000-03-17T02:59:20.639Z'
      }
    })

  t.is(response.status, 200)

  t.deepEqual(response.body.id, 90001)
  t.deepEqual(response.body.auth0Id, 60926)
  t.deepEqual(response.body.locale, 'en_BORK')
  t.deepEqual(response.body.nickname, 'neural')
  t.deepEqual(response.body.status, 'magenta Personal Loan Account')
  t.deepEqual(
    response.body.avatar,
    'https://s3.amazonaws.com/uifaces/faces/twitter/koridhandy/128.jpg'
  )
  t.deepEqual(response.body.created_at, '2000-05-02T18:39:05.087Z')
  t.deepEqual(response.body.updated_at, '2000-03-17T02:59:20.639Z')
})
test('update should update user', async (t) => {
  const {request} = t.context

  // mock jwks
  const token = createToken(privateKey, '123', currentUser)
  jwksEndpoint('http://localhost', [{pub: publicKey, kid: '123'}])

  const response = await request
    .put('/api/users/19270')
    .set('Authorization', \`Bearer \${token}\`)
    .send({
      article: {
        id: 19270,
        auth0Id: 60926,
        locale: 'en_BORK',
        nickname: 'neural',
        status: 'magenta Personal Loan Account',
        avatar:
          'https://s3.amazonaws.com/uifaces/faces/twitter/koridhandy/128.jpg',
        created_at: '2000-05-02T18:39:05.087Z',
        updated_at: '2000-03-17T02:59:20.639Z'
      }
    })

  t.is(response.status, 200)

  t.deepEqual(response.body.id, 19270)
  t.deepEqual(response.body.auth0Id, 60926)
  t.deepEqual(response.body.locale, 'en_BORK')
  t.deepEqual(response.body.nickname, 'neural')
  t.deepEqual(response.body.status, 'magenta Personal Loan Account')
  t.deepEqual(
    response.body.avatar,
    'https://s3.amazonaws.com/uifaces/faces/twitter/koridhandy/128.jpg'
  )
  t.deepEqual(response.body.created_at, '2000-05-02T18:39:05.087Z')
  t.deepEqual(response.body.updated_at, '2000-03-17T02:59:20.639Z')
})
test('delete should delete user', async (t) => {
  const {request, Article} = t.context

  let articles = await Article.query()
  t.deepEqual(articles.length, 3)

  // mock jwks
  const token = createToken(privateKey, '123', currentUser)
  jwksEndpoint('http://localhost', [{pub: publicKey, kid: '123'}])

  const response = await request
    .delete('/api/users/19270')
    .set('Authorization', \`Bearer \${token}\`)

  articles = await Article.query()
  t.deepEqual(articles.length, 2)

  t.is(response.status, 204)
  t.deepEqual(response.body, {})
})

/* mat Custom tests [start] */
/* mat Custom tests [end] */
"
`;
